### DISCUSSION TITLE
Is it possible to use ProviderScope.overrides for AsyncNotifierProviderFamily?

### DISCUSSION THREAD

**Discussion by enm10k** [rrousselGit/riverpod#2024] [Q&A] [answered]
To retrieve the state from AsyncNotifierProviderFamily, the family key is required.
However, it can be cumbersome to pass the family key every time, especially when the widget tree is nested deeply.

Is it possible to use ProviderScope.overrides for this issue like @rrousselGit mentioned [here](https://mobile.twitter.com/remi_rousselet/status/1278545540490412032)?

>I f this is about "I don't want to pass the Family key all around the widget tree", a solution is to create an extra provider that exposes the combination of the family + the key
>
> This can be combined with `ProviderScope.overrides` for fancy stuff
>
> [https://gist.github.com/rrousselGit/ddd92ae211bee94b0188ddec4e01e2cd](https://t.co/Dkq7YForRB)

**Comment by rrousselGit** [comment, marked as answer]
Are you looking for `overrides: [provider.overrideWIth(...)]`?

**Reply by enm10k** [reply]
Thanks for your reply.

Could you give me some advice on how to use overrideWith?
I tried the following code, but I couldn't understand how to pass the family key.

```
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';

final sampleFamily = AsyncNotifierProviderFamily<SampleNotifier, int, int>(
  () => SampleNotifier(),
);

class SampleNotifier extends FamilyAsyncNotifier<int, int> {
  @override
  FutureOr<int> build(int arg) {
    // In the actual code, I'm calling an API here.
    return Future.value(3 * arg);
  }
}

void main() {
  sampleFamily.overrideWith(() => SampleNotifier());
  runApp(
    ProviderScope(
      overrides: [
        // I tried to use overrideWith, but I couldn't pass the family key.
        sampleFamily.overrideWith(() => SampleNotifier()),
      ],
      child: const App(),
    ),
  );
}

class App extends HookConsumerWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final value = ref.watch(sampleFamily(3));

    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: Text(value.toString()),
        ),
      ),
    );
  }
}
```

**Reply by rrousselGit** [reply]
What do you mean by "couldn't pass the family key"? What are you trying to pass, where? The AsyncNotifier should already receive the family key

**Reply by enm10k** [reply]
I was expecting to be able to use the provider without specifying the family key in the App class, by specifying the family key using ProvierScope.overrides.

**Reply by rrousselGit** [reply]
That's not a thing. You always have to specify the family key

**Reply by enm10k** [reply]
Thank you for answering.

This is just a question and not a request, but is it possible to add a feature like the one I described?
Do you think it is worth considering?

**Reply by rrousselGit** [reply]
It's possible, yes. There's an issue about it.
I don't like the idea though. Scoping is a bit "advanced", and it'd make refactoring from a provider to a family weird.

Instead you could make a separate provider used to obtain the family key

Like:

```dart
final provider = Provider<int>((ref) => throw unimplemented());
...
overrides: [provider.overrideWithValue(familyKey)])
...

final id = ref.watch(provider);
final value = ref.watch(family(id));
```
